lib/jekyll/document.rb

```diff
@@ -426,7 +426,7 @@ def merge_categories!(other)
 
     private
     def merge_date!(source)
-      if data.key?("date") && !data["date"].is_a?(Time)
+      if data.key?("date")
```

Comment:
I see the addition of tests :+1:

What’s this line about?

Reply:
Sorry, didn't had time yesterday to properly update this PR with more information.

This change actually fixes the inconsistent behaviour, because as long as `data['date']` is set it will call [`Utils.parse_date`](https://github.com/jekyll/jekyll/blob/603f2ea52f52cd3a64e37371349d8292f3a83655/lib/jekyll/utils.rb#L135-L139) which transforms the given data into `localtime` (as determined by `ENV['TZ']`). This is exactly the behaviour that SafeYAML was doing automatically before Ruby 2.4.
Because of the old version of this line there were no inconsistencies between the different Ruby versions when `data['date']` was interpreted as a `String` rather than a `Time`.

That said, I am not sure the behaviour of Ruby 2.3 is my preferred one, i would actually prefer the behaviour of newer ruby versions but the most important thing is to fix the incosistency and i would argue that jekyll users are more used to the behaviour of `everything gets converted into whatever ENV['TZ'] is`.